File: backend/app/schemas/progress.py
from pydantic import BaseModel
from datetime import datetime


class OverallProgress(BaseModel):
    total_concepts: int
    concepts_started: int
    concepts_mastered: int
    total_questions_answered: int
    overall_accuracy: float
    current_streak: int
    longest_streak: int
    total_xp: int
    mastery_distribution: dict[str, int]  # {"novice": 10, "learning": 5, ...}


class TopicProgress(BaseModel):
    topic_id: str
    topic_name: str
    total_concepts: int
    mastered_concepts: int
    avg_confidence: float
    mastery_percent: float


class ConceptProgressDetail(BaseModel):
    concept_id: str
    concept_name: str
    topic_name: str
    mastery_level: str
    confidence_score: float
    exposure_count: int
    correct_count: int
    incorrect_count: int
    accuracy: float
    last_seen_at: datetime | None
    next_review_at: datetime | None
    error_streak: int


class WeakAreaResponse(BaseModel):
    concept_id: str
    concept_name: str
    topic_name: str
    confidence_score: float
    error_streak: int
    accuracy: float
    recommended_action: str


class DailyStatsResponse(BaseModel):
    date: str
    questions_answered: int
    correct_count: int
    accuracy: float
    xp_earned: int
    time_spent_minutes: float


class StreakResponse(BaseModel):
    current_streak: int
    longest_streak: int
    today_completed: bool
    daily_goal: int
    questions_today: int

File: frontend/src/api/progress.ts
import api from './client';

export interface OverallProgress {
  total_concepts: number;
  concepts_started: number;
  concepts_mastered: number;
  total_questions_answered: number;
  overall_accuracy: number;
  current_streak: number;
  longest_streak: number;
  total_xp: number;
  mastery_distribution: Record<string, number>;
}

export interface TopicProgress {
  topic_id: string;
  topic_name: string;
  total_concepts: number;
  mastered_concepts: number;
  avg_confidence: number;
  mastery_percent: number;
}

export interface WeakArea {
  concept_id: string;
  concept_name: string;
  topic_name: string;
  confidence_score: number;
  error_streak: number;
  accuracy: number;
  recommended_action: string;
}

export interface DailyStats {
  date: string;
  questions_answered: number;
  correct_count: number;
  accuracy: number;
  xp_earned: number;
  time_spent_minutes: number;
}

export interface StreakInfo {
  current_streak: number;
  longest_streak: number;
  today_completed: boolean;
  daily_goal: number;
  questions_today: number;
  xp_today: number;
}

export const progressApi = {
  getOverview: (subjectId: string) =>
    api.get<OverallProgress>(`/progress/overview/${subjectId}`),

  getTopicProgress: (subjectId: string) =>
    api.get<TopicProgress[]>(`/progress/topics/${subjectId}`),

  getWeakAreas: (subjectId: string) =>
    api.get<{ items: WeakArea[]; total: number; page: number; page_size: number }>(`/progress/weak-areas/${subjectId}?page_size=100`),

  getDailyStats: (days = 30) =>
    api.get<{ items: DailyStats[]; total: number }>(`/progress/daily-stats?days=${days}`),

  getStreak: () =>
    api.get<StreakInfo>('/progress/streak'),
};
